Sources/Fuzzilli/FuzzIL/ProgramTypes.swift

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Store known types for program variables at specific instructions
+public struct TypeIndexPair: Equatable {
+    public let index: Int
+    public let type: Type
+
+    public init(index: Int, type: Type){
+        self.index = index
+        self.type = type
+    }
+
+    public static func == (lhs: TypeIndexPair, rhs: TypeIndexPair) -> Bool {
+        return lhs.index == rhs.index && lhs.type == rhs.type
+    }
+}
+
+public struct ProgramTypes: Equatable, Sequence {
+    // types[variable][instructionIndex] = type
+    private var types = VariableMap<[TypeIndexPair]>()
+
+    public init () {}
+
+    public init (from types: VariableMap<[TypeIndexPair]>) {
+        self.types = types
+    }
+
+    public static func == (lhs: ProgramTypes, rhs: ProgramTypes) -> Bool {
+        return lhs.types == rhs.types
+    }
+
+    // Save type of given variable
+    public mutating func setType(of variable: Variable, to type: Type, at instrIndex: Int) {
+        // Initialize type structure for given variable if not already
+        if types[variable] == nil {
+            types[variable] = []
+        }
+        let typeIndexPair = TypeIndexPair(index: instrIndex, type: type)
+        // Check if we update type of instruction in the middle
+        if let insertionIndex = types[variable]!.firstIndex(where: { el in el.index >= instrIndex }) {
+            if types[variable]![insertionIndex].index == instrIndex {
+                // Overwrite old type
+                types[variable]![insertionIndex] = typeIndexPair
+            } else {
+                // Add new type information at instrIndex
+                types[variable]!.insert(typeIndexPair, at: insertionIndex)
+            }
+        } else {
+            types[variable]!.append(typeIndexPair)
+        }
+    }
+
+    // Get type of variable at current instruction
+    public func getType(of variable: Variable, at instrIndex: Int) -> Type {
+        if let instrTypes = types[variable],
+           let typeData = instrTypes.last(where: { el in el.index <= instrIndex })
+        {
+            return typeData.type
+        }
+
+        return .unknown
```

Comment:
I wonder if there is something else we should do here in the case where the variable exists, and we have type information for it, but isn't yet defined at the queried index (which would probably be a bug in the user code). We could return `.nothing`, but maybe we should for now just assert that that case doesn't happen (I'm just slightly worried that simply returning `.unknown` here might hide some bugs in the future).

Reply:
hmm but how do we decide whether variable does not exist at all or we just don't have data for it? This should not be responsible of this struct. If we want some additional checks related to program, it should be wrapped by Program API

Reply:
I was thinking about something along the lines of
```
// Currently, we disallow querying the type of a variable before it's defined
assert(types[variable] == nil);
```
But yes, that assumes that we will either have no type at all or at least the type at the point of definition (which is probably sensible?). But I'm also fine with adding such a check later in the Program API.

Reply:
well in this struct we store also runtime types and for example we don't gather function types, literal types, ... at all

Reply:
Yeah, but for those `types[variable]` will be `nil`, no?
As far as I understand, we either have `types[variable] == nil` if we don't have any information, or we will have an array of at least size 1, and the first entry contains the type at the definition of the variable. So the assert would work to ensure that users don't query types of yet-undefined variables.

Reply:
There is no explicit rule if there is some type info then there is type on definition (and this struct should not rely on it).
For example we don't collect types on literal definitions, but there are instructions that can change them using side-effect (and in future we can somehow collect them)
```
const v0 = "a"
const v1 = v0++
```

So `We disallow querying the type of a variable before it's defined` would not be true.
Rather `Currently, we disallow querying the type of a variable before we have some type info.`

And still I think this struct should not care about variable definition and other stuff, this should be just storage.
Also these not-explicit rules are good source of bugs as well (somebody changes TypeCollectionAnalyzer and this storage will break, and we cannot enforce it by unit tests as it involves JS execution)

Reply:
Ok makes sense!